### PR TITLE
resolve  Check whether special is active by tag name or by id

### PR DESCRIPTION
**Pull request by newfrenchy83**
resolve https://github.com/wesnoth/wesnoth/issues/3915 with add of special_id and special_tags( and idem for special_active)

### CODE REVIEW

**Comment by newfrenchy83** [comment]
@jostephd could you test this PR please?

**Comment by gfgtdf** [comment]
I didn't look at it in detail, but it seems to me that there is quite some space for refactoring/ removing code dublication.

**Comment by newfrenchy83** [comment]
i remove code duplication.

**Comment by newfrenchy83** [comment]
@gfgtdf  code duplication corrected and sorry for the late.

**Comment by CelticMinstrel** [comment]
This seems kind of strange. Why do we now have three different keys? We have `special`, `special_id`, and `special_tags`. Aren't there only two ways to test a special? It can match by tag or by ID, right? Doesn't `special` already go by tag? I see its code _is_ slightly different from that for `special_tags`, but is there any actual semantic difference?

**Comment by newfrenchy83** [comment]
If i filter drains special with special key then the filter weapon with drains id but also with drains tags, with special_id i match id=drains only but not the orher drains tags weapon. But yes i can remove special_tags and use what special_id and special.

**Comment by jostephd** [comment]
I think what happens is that special=foo tries to match foo both as an id and as a tag name, and then special_id=bar and special_tags=baz try to match bar only as an id and baz only as a tag. In which case the right thing is probably to support all three but mark special=foo as deprecated (https://wiki.wesnoth.org/CompatibilityStandards).

**Comment by newfrenchy83** [comment]
I am of the same opinion as @jostephd concerning special = foo, and it is better to have special_id on the one hand and special_tags on the other hand to avoid filtering a type of special attack when we only target only one in this type.

**Comment by CelticMinstrel** [comment]
For consistency with ability filtering in [StandardUnitFilter](https://wiki.wesnoth.org/StandardUnitFilter), I'd say you should have `special=` that matches on ID and `special_type` that matches on tag name. Unfortunately that's problematic with respect to backwards compatibility... but you can at least rename your `special_tags` to `special_type`.

**Comment by CelticMinstrel** [comment]
I probably should've mentioned this explicitly before, but... again for consistency, please use `special_type_active` instead of `special_active_type`. Similarly, I'd prefer `special_id_active` instead of `special_active_id`.

**Comment by CelticMinstrel** [comment]
Just wondering, regarding deprecation... would it be acceptable to deprecate `special` and `special_active` for now, and after a certain period, change its behaviour to match `special_id` and `special_id_active`? Then, perhaps after another period of time, undeprecate the first two and deprecate the second two.

**Comment by newfrenchy83** [comment]
@CelticMinstrel This would mean that special_id should be invalidated after inserting it into the wml code with special_type so that special, at that time, , but at the same time modify the behavior of special would prevent them from reworking the code, view that all or almost use special to refer to the id of the special weapon and not the type, in this respect yes, your proposal is acceptable to use special_id transiently .

**Comment by newfrenchy83** [comment]
@CelticMinstrel have you forget this PR?

**Comment by CelticMinstrel** [comment]
I did, actually.

I don't see any problems to prevent this being merged at this point.

**Comment by newfrenchy83** [comment]
@CelticMinstrel this PR will be commited a day?

**Comment by CelticMinstrel** [comment]
I think it can probably be merged, yes. I'd rather not be the one to do it though, unless there's several people here saying the same thing.

@jostephd @gfgtdf - what do you think?

**Comment by jostephd** [comment]
I don't have time to review this currently, but I'll take a look when I have time for wesnoth again.

**Comment by newfrenchy83** [comment]
@CelticMinstrel 2 things: primo i covnerted special for filter id only and remove special_id. secundo nor @jostephd or @gfgtdf have review this PR.

**Comment by jostephd** [comment]
Set milestone.

**Comment by newfrenchy83** [comment]
Listen, i know what you're many PR to review, but please don't forget this PR.

**Comment by Wedge009** [comment]
Just looking at the code, it seems to make sense to me. I'll try to test when I get home - what units make use of weapon specials? Actually, never mind, I just saw #3915 referring to the first strike ability in HttT.

@CelticMinstrel Any particular reason for the reticence to merge or just preferring to let @jostephd check first?

**Comment by CelticMinstrel** [comment]
I think my reticence to merge was solely that I didn't have the time to test it. The code looks okay from here, I think.

**Comment by Wedge009** [comment]
I'm not really sure what I'm checking here, but I note Li'sar is the only one with the campaign-specific initiative special.

...after staring at this for a while I think I'm not looking in the right place. I would need to find how the tag (firststrike) and ids (initiative) can be inspected/filtered.

To add to the confusion, I found that the Wesnoth French translation for first strike is also initiative. (first strike is a core special) XD

**Comment by Wedge009** [comment]
Does this affect/help #1123 or is that one referring to the translatable name?

**Comment by newfrenchy83** [comment]
This PR don't concern https://github.com/wesnoth/wesnoth/issues/1123 for what i know, but weapon and abilities like weapon( drains by example because id and tags are both), but the core has two example only(Li'sar firstrike abilities and utbs formation ability), and added abilities no used in the core don't good idea.
[abilities_weapons.cfg.txt](https://github.com/wesnoth/wesnoth/files/3529250/abilities_weapons.cfg.txt)
this files must be help you to understand.

**Comment by Wedge009** [comment]
eg tag 'slow' differs from id 'slow_leadership' - and filtering for either doesn't work in current code but is possible in this PR?

**Comment by newfrenchy83** [comment]
Slow abilitiz havenir filter, where filter din woek ?

**Comment by newfrenchy83** [comment]
You use ability or special for forum ltering slow_leadership becaise if you use special then this don't work in code actual or this Pr because slow_leadership is a abilities nor an weapon stricto sensu and must be lacef in abilities tags. You must use abilitie filter for filtering the abilities

**Comment by jostephd** [comment]
@Wedge009 #1123 is about showing the ability's name in the UI. This issue is about being able to filter for `[leadership] id=inspires` using either `tag="leadership"` or `id="inspires"`. In master the code takes just one string argument and compares it both to the tag and to the id.

**Comment by newfrenchy83** [comment]
i put a test case

**Comment by CelticMinstrel** [comment]
Was the schema change actually necessary? I mean, was schema validation actually failing? I'm only asking because I thought I remembered making that update already (maybe I forgot to push it or something).

Also, if I recall correctly, you could have instead set `name="firststrike"` to be a comma-separated list of the various tag names, like `name="firststrike,petrifies"` etc. Assuming that _does_ work I would definitely prefer it.

**Comment by newfrenchy83** [comment]
with drains abilities in scenario-test, i must include all ablities in shema validation. for comma-separated list of the various tag names, like name="firststrike,petrifies" etc., i don't know but i will study this.

**Comment by newfrenchy83** [comment]
yes shema validation failing actually because scenario-test.

**Comment by CelticMinstrel** [comment]
Those `SIMPLE_KEY` additions might not be needed. Instead changing the `super` key to contain `~value~` instead of `~generic~`.

**Comment by jostephd** [comment]
@newfrenchy83 Please don't close/reopen issues, that generates notifications to everyone who follows the issue.

**Comment by CelticMinstrel** [comment]
Can you explain your change to the `BASED_ON_SPECIAL` macro? It doesn't look correct to me.

**Comment by newfrenchy83** [comment]
@CelticMinstrel Just because when I inserted drains ability in the scenario-test file, the change was not validated, perhaps because the BASED_ON_SPECIAL macro had to validate special attacks and not similar abilities but placed in abilities tags; I modified the macro accordingly for this to work.

**Comment by CelticMinstrel** [comment]
The entire point of `BASED_ON_SPECIAL` is to be able to reuse the weapon special definition when defining the aura ability based on that weapon special. As far as I can tell, your new version no longer does this - the `super` key has been changed to reference an ability rather than a weapon special.

The added `FILTER_TAG`s _might_ be correct if they are not present in the weapon special definition.

**Comment by newfrenchy83** [comment]
The super key change has changed to reference an ability because this is an ability similar to weapon specials, i will trying to reuse original macro but i doubt of result.

**Comment by newfrenchy83** [comment]
https://github.com/CelticMinstrel it is fixed

File: data/schema/units/abilities.cfg [hunk 1]
```diff
@@ -4,7 +4,21 @@
...
+		{SIMPLE_KEY affect_allies bool}
+		{SIMPLE_KEY affect_enemies bool}
+        [tag]
+		name="affect_adjacent"
+		[key]
+			name="adjacent"
+			type="dir_list"
+		[/key]
+		{FILTER_TAG "filter" unit ()}
+	[/tag]
```
Comment (CelticMinstrel) [inline data/schema/units/abilities.cfg]: I think I'm seeing what the issue here is now - we probably need multiple inheritance here. Try replacing everything from line 6 to 21 with just this line:

```
super="units/unit_type/abilities/~generic~,units/unit_type/attack/specials/" + {NAME}
```

**Comment by newfrenchy83** [comment]
@CelticMinstrel i'm apply change but i must let filter_weapon and filter_second_weapon tags in macro.
i pull https://github.com/wesnoth/wesnoth/pull/4357 for what shema validation could be commited in a separate way of this PR.

**Comment by newfrenchy83** [comment]
Because shema validation in https://github.com/wesnoth/wesnoth/pull/4373, i must erase testvaliadion or create another PR for replace that there.

**Pull request closed without merging**